src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs

```diff
@@ -18,6 +19,25 @@ public static class CollectionsMarshal
         public static Span<T> AsSpan<T>(List<T>? list)
             => list is null ? default : new Span<T>(list._items, 0, list._size);
 
+        /// <summary>
+        /// Gets a ref to a <typeparamref name="TValue"/> in the <see cref="Dictionary{TKey, TValue}"/>.
+        /// </summary>
+        /// <param name="dictionary">The dictionary to get the ref to <typeparamref name="TValue"/> from.</param>
+        /// <param name="key">The key used for lookup.</param>
+        /// <remarks>Items should not be added or removed from the <see cref="Dictionary{TKey, TValue}"/> while the ref <typeparamref name="TValue"/> is in use.</remarks>
+        /// <exception cref="KeyNotFoundException">Thrown when <paramref name="key"/> does not exist in the <paramref name="dictionary"/>.</exception>
+        public static ref TValue GetValueRef<TKey, TValue>(Dictionary<TKey, TValue> dictionary, TKey key) where TKey : notnull
+        {
+            ref TValue valueRef = ref dictionary.FindValue(key);
+
+            if (Unsafe.IsNullRef(ref valueRef))
+            {
+                ThrowHelper.ThrowKeyNotFoundException(key);
+            }
+
+            return ref valueRef;
+        }
```

Comment:
The implementation of the approved API looks good.

That said, I still question who's actually going to use this method.  If you care about this level of performance optimization, in what situation are you going to want it to throw an exception if the key can't be found?

Are there any places in all of dotnet/runtime where a) the performance benefits of this method will be meaningful and b) we want the exception behavior when the key isn't found?

Reply:
> "[...] I still question who's actually going to use this method. If you care about this level of performance optimization, in what situation are you going to want it to throw an exception if the key can't be found? i.e. why wouldn't you use GetValueRefOrNullRef?"

I can agree this is likely the least useful API of the 3 approved in this group, yeah. I'm thinking a valid use case could be the same as `GetValueRefOrNullRef` (as in, where you'd use this method for the same reasons why you'd use `GetValueRefOrNullRef` over just the standard indexer or `ContainsKey`/`TryGetValue` + indexer), except in cases where you never expect the key not to be present in the dictionary at that point, so you'd just use this one to avoid doing the check on your own? The two would be functionally equivalent, yeah, with the only difference just being that this would make the user code less verbose in this scenario 🤔

Reply:
Thanks.  Can you find a real piece of code somewhere where this would make a meaningful difference? It would need to be something where only part of the fetched value is needed, the overhead of copying the whole value unnecessarily is measurable, an exception on failure is desired to the point where on this hot path the extra check required for the implementation to throw on null is required, a not found exception rather than a null reference exception if not found is important, etc. I'm struggling to think of any situation where I'd want to use this method over one of the others.

Reply:
I'll admit I haven't been able to find applicable code snippets myself (looked in `System.Private.CoreLib`, `ImageSharp`, and tried asking on Discord), so at this time I'm not able to provide much evidence to argue for this API other than the general use case scenario support that @benaadams already mentioned in the original proposal in #27062. As mentioned, I can definitely agree that this is the least useful API of the three as it doesn't offer new functionality per se but it's more like a convenience wrapper (personally I mostly just care about #54611 making it to .NET 6 😄).

Let me know how/if you'd like to proceed with this, I guess we could either just merge it given that it's a very minor change anyway and the API itself has already been approved by FXDC, or we could just close this for now 🙂

Reply:
My preference is we close this PR and "unapprove" the API; we shouldn't add it just because we can... new surface area needs good justification.  @dotnet/area-system-collections can make a call, though.

Reply:
@stephentoub you are saying unapprove only `GetValueRef` which throws, correct?

I'd agree that this is mostly just a "convenience" API so users don't need to deal with `NullRef` and handle the logic themselves, it doesn't represent anything that they can't do themselves. Given the other considerations around the CollectionMarshal APIs, forcing users to understand `NullRef` also seems the least of the worries 😄

Reply:
> you are saying unapprove only GetValueRef which throws, correct?

Correct

Reply:
I'd personally be fine with ditching the throwing variant.

Reply:
I'll go ahead and close this then since we all agree this API in particular is not strictly necessary.
There's only #54611 left to review and merge then 😄